streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -375,68 +410,77 @@ public void queryOnRebalance() throws Exception {
+                    DEFAULT_TIMEOUT_MS);
                 verifyAllWindowedKeys(
-                    streamRunnables,
-                    streamRunnables[i].getStream(),
-                    streamRunnables[i].getStateListener(),
+                    streamsList,
+                    streamsList.get(i),
+                    listeners.get(i),
                     inputValuesKeys,
                     windowStoreName + "-" + streamThree,
                     0L,
-                    WINDOW_SIZE);
-                assertEquals(KafkaStreams.State.RUNNING, streamRunnables[i].getStream().state());
+                    WINDOW_SIZE,
+                    DEFAULT_TIMEOUT_MS);
             }
 
             // kill N-1 threads
-            for (int i = 1; i < numThreads; i++) {
-                streamRunnables[i].close();
-                streamThreads[i].interrupt();
-                streamThreads[i].join();
+            for (int i = 1; i < streamsList.size(); i++) {
+                final Duration closeTimeout = Duration.ofSeconds(60);
+                assertThat(String.format("Streams instance %s did not close in %d ms", i, closeTimeout.toMillis()),
+                    streamsList.get(i).close(closeTimeout));
             }
 
+            waitForApplicationState(streamsList.subList(1, numThreads), State.NOT_RUNNING, Duration.ofSeconds(60));
+
+            // Even though the closed instance(s) are now in NOT_RUNNING there is no guarantee that
+            // the running instance is aware of this, so we must run our follow up queries with
+            // enough time for the shutdown to be detected.
+
             // query from the remaining thread
             verifyAllKVKeys(
-                streamRunnables,
-                streamRunnables[0].getStream(),
-                streamRunnables[0].getStateListener(),
+                streamsList,
+                streamsList.get(0),
+                listeners.get(0),
                 inputValuesKeys,
-                storeName + "-" + streamThree);
+                storeName + "-" + streamThree,
+                DEFAULT_TIMEOUT_MS);
             verifyAllWindowedKeys(
-                streamRunnables,
-                streamRunnables[0].getStream(),
-                streamRunnables[0].getStateListener(),
+                streamsList,
+                streamsList.get(0),
+                listeners.get(0),
                 inputValuesKeys,
                 windowStoreName + "-" + streamThree,
                 0L,
-                WINDOW_SIZE);
-            assertEquals(KafkaStreams.State.RUNNING, streamRunnables[0].getStream().state());
+                WINDOW_SIZE,
+                DEFAULT_TIMEOUT_MS);
+
+            // It's not enough to assert that the first instance is RUNNING because it is possible
+            // for the above checks to succeed while the instance is in a REBALANCING state.
+            waitForApplicationState(streamsList.subList(0, 1), State.RUNNING, Duration.ofSeconds(60));
```

Comment:
Why we wait for state after the verification than before it?

Reply:
Only because that is the way it was in the original test. It also turned up the interesting fact that the stores are queryable before the instance is in a RUNNING state. It seems OK to be in either order provided that we're giving enough time to get to a running state, but I can re-arrange them if that is needed to get this committed.

Reply:
Since we merged in KIP-429, under normal processing the state of streams would be in REBALANCING for very short time in between poll calls only during restoration, I think that's why you observed that it is queryable, as the state transition to RUNNING happens instantaneously after line 454 (i.e. the restoration takes no time).

Just for the logical reasoning itself, I think it is still better to check state transition first, and then verify queryable results.